File: sdks/python/openforest_sdk/stdio.py
from __future__ import annotations

import json
import secrets
import sys
from typing import Any, Callable

from .commit import commit_hash


BotFn = Callable[[dict[str, Any]], list[dict[str, Any]]]


def run_stdio(bot_fn: BotFn) -> None:
    pending: dict[int, tuple[list[dict[str, Any]], str]] = {}
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        message = json.loads(line)
        msg_type = message.get("type")
        tick = message.get("tick")
        if msg_type == "commit":
            observation = message.get("observation", {})
            actions = bot_fn(observation)
            nonce = secrets.token_hex(8)
            pending[int(tick)] = (actions, nonce)
            commit = commit_hash(actions, nonce)
            response = {"type": "commit", "tick": tick, "commit": commit}
            sys.stdout.write(json.dumps(response) + "\n")
            sys.stdout.flush()
        elif msg_type == "reveal":
            actions, nonce = pending.pop(int(tick), ([], ""))
            response = {"type": "reveal", "tick": tick, "actions": actions, "nonce": nonce}
            sys.stdout.write(json.dumps(response) + "\n")
            sys.stdout.flush()

File: sdks/node/src/stdio.ts
import readline from "readline";
import crypto from "crypto";
import { commitHash } from "./commit";

export type BotFn = (observation: Record<string, unknown>) => Array<Record<string, unknown>>;

export function runStdio(botFn: BotFn): void {
  const pending = new Map<number, { actions: Array<Record<string, unknown>>; nonce: string }>();
  const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });

  rl.on("line", (line) => {
    if (!line.trim()) {
      return;
    }
    const message = JSON.parse(line);
    const msgType = message.type;
    const tick = Number(message.tick ?? 0);

    if (msgType === "commit") {
      const observation = message.observation ?? {};
      const actions = botFn(observation);
      const nonce = crypto.randomBytes(8).toString("hex");
      pending.set(tick, { actions, nonce });
      const commit = commitHash(actions, nonce);
      const response = { type: "commit", tick, commit };
      process.stdout.write(JSON.stringify(response) + "\n");
    } else if (msgType === "reveal") {
      const stored = pending.get(tick) ?? { actions: [], nonce: "" };
      pending.delete(tick);
      const response = { type: "reveal", tick, actions: stored.actions, nonce: stored.nonce };
      process.stdout.write(JSON.stringify(response) + "\n");
    }
  });
}
